### PR TITLE
[libc++][test] Improve tests for assign in std::vector and vector<bool>

### PR DESCRIPTION
**Pull request by winner245**
This PR enhances the test coverage for the `assign` function of `std::vector` by adding new tests for several important test cases that were previously missing, as shown in the following table:

| test cases | forward_iterator | input_iterator |
|----------|----------|----------|
| new_size > capacity()   |  Yes  | Yes |
| size() < new_size <= capacity()   | No   | No   |
| new_size <= size()    | No   | No   | 

Similarly, no tests have previously covered `assign(_InputIterator __first, _InputIterator __last)` and `assign(size_type __n, const value_type& __x)` for  `vector<bool>`.

With this patch applied, all missing tests will be covered.

### CODE REVIEW

File: libcxx/test/std/containers/sequences/vector/vector.cons/assign_iter_iter.pass.cpp [hunk 1]
```diff
@@ -63,6 +62,100 @@ TEST_CONSTEXPR_CXX20 bool test() {
...
+      for (std::size_t i = 0; i < 2; ++i)
+        v.emplace_back(99);
+      v.assign(It(arr2), It(std::end(arr2)));
+      assert(v.size() == 3);
+      assert(v[0].value == 1);
+      assert(v[1].value == 101);
+      assert(v[2].value == 42);
+    }
+  }
+  { // Test with size() < new_size < capacity() for inputs_iterator, resulting in construction at end during assign
```
Comment (ldionne) [inline libcxx/test/std/containers/sequences/vector/vector.cons/assign_iter_iter.pass.cpp:137]:
```suggestion
  { // Test with size() < new_size < capacity() for input_iterator, resulting in construction at end during assign
```

File: libcxx/test/std/containers/sequences/vector/vector.cons/assign_iter_iter.pass.cpp
Comment (ldionne) [inline libcxx/test/std/containers/sequences/vector/vector.cons/assign_iter_iter.pass.cpp:1]: Would it make sense to add the same tests to `vector<bool>`?

**Review by ldionne — COMMENTED** [review]
Great catch! I played around and we indeed didn't catch bugs in the code paths you described, which is quite bad. Thanks for adding coverage!

File: libcxx/test/std/containers/sequences/vector/vector.cons/assign_iter_iter.pass.cpp
Comment (winner245) [reply libcxx/test/std/containers/sequences/vector/vector.cons/assign_iter_iter.pass.cpp:1]: Sure. I just checked and it looks like the tests for `vector<bool>::assign` are completely missing. I will add more tests.

File: libcxx/test/std/containers/sequences/vector.bool/assign_iter_iter.pass.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// <vector>
+
+// void assign(_InputIterator __first, _InputIterator __last);
+// void assign(_ForwardIterator __first, _ForwardIterator __last);
```
Comment (ldionne) [inline libcxx/test/std/containers/sequences/vector.bool/assign_iter_iter.pass.cpp:12]:
```suggestion
// template <class InputIter>
// void assign(InputIter first, InputIter last);
```

File: libcxx/test/std/containers/sequences/vector.bool/assign_size_value.pass.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// <vector>
+
+// void assign(size_type __n, const value_type& __x);
```
Comment (ldionne) [inline libcxx/test/std/containers/sequences/vector.bool/assign_size_value.pass.cpp:11]:
```suggestion
// void assign(size_type n, const value_type& x);
```

**Review by ldionne — APPROVED** [review]
LGTM, thanks! Please rebase onto `main` (you'll get a small merge conflict in `vector::assign` tests because of a change I made, sorry about that).

File: libcxx/test/std/containers/sequences/vector.bool/assign_iter_iter.pass.cpp [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+      std::vector<bool> in(128, true);
+      std::vector<bool> v(5, false);
+      assert(v.capacity() < in.size());
+      using It = forward_iterator<std::vector<bool>::iterator>;
+      v.assign(It(in.begin()), It(in.end()));
+      assert(v == in);
+    }
+    { // No reallocation: fit wintin current size
+      bool in[]                    = {false, true, false, true, true};
+      TEST_CONSTEXPR std::size_t N = sizeof(in) / sizeof(in[0]);
```
Comment (ldionne) [inline libcxx/test/std/containers/sequences/vector.bool/assign_iter_iter.pass.cpp:31]: This doesn't need to be constexpr so I'd leave it out. We use `TEST_CONSTEXPR` only when it's really required for the test to work.

File: libcxx/test/std/containers/sequences/vector.bool/assign_size_value.pass.cpp [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    { // Reallocation happens
+      std::size_t N = 128;
+      std::vector<bool> v(5, false);
+      assert(v.capacity() < N);
+      v.assign(N, true);
+      assert(v.size() == N);
+      for (std::size_t i = 0; i < N; ++i)
+        assert(v[i] == true);
+    }
+    { // No reallocation: fit wintin current size
```
Comment (ldionne) [inline libcxx/test/std/containers/sequences/vector.bool/assign_size_value.pass.cpp:29]:
```suggestion
    { // No reallocation: fit within current size
```

File: libcxx/test/std/containers/sequences/vector.bool/assign_size_value.pass.cpp [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+      std::size_t N = 5;
+      std::vector<bool> v(2 * N, false);
+      v.assign(N, true);
+      assert(v.size() == N);
+      for (std::size_t i = 0; i < N; ++i)
+        assert(v[i] == true);
+    }
+    { // No reallocation: fit wintin spare space
```
Comment (ldionne) [inline libcxx/test/std/containers/sequences/vector.bool/assign_size_value.pass.cpp:37]:
```suggestion
    { // No reallocation: fit within spare space
```

**Review by ldionne — COMMENTED** [review]
Sorry for the many small reviews, I didn't do a good job looking at this patch the first time around and I found a few more nitpicks. This is great though, they're all minor comments.

**Review by ldionne — APPROVED** [review]
Okay, now this should be good to go for real once CI is green. Sorry for the multiple notifications. Thanks a lot for the patch!

**Comment by winner245** [comment]
Thank you so much for your careful and thorough review! I truly appreciate the time and effort you've put into helping improve all my PRs. Your careful review has identified all my careless points and significantly improved my work. Please don't worry about the number of comments; they are indeed incredibly helpful! I'll go through all of them and make the necessary adjustments.

**Review by ldionne — APPROVED** [review]
LGTM with green CI. I merged `main` into your branch to get over a CI issue that was present in `main`.

**Pull request merged**